### PR TITLE
Feature/937 & 938 Waiting for Review and Previous Reviews on Dashboard

### PR DESCRIPTION
**Pull request by emullaney**
#937 Waiting for Review block on Dashboard
**Acceptance criteria - as an OTF staff member**

- Title of block is 'Waiting for your review'
- Number of applications I am assigned to review displays alongside title as per design
- Lists applications I am assigned to review
- Applications ordered by submission date descending
- Table columns should be as per All Submissions tables
- Limit number of submissions to 5, if more than 5 show all link to appear (as per other summary tables) and link to all submissions page pre-filtered by reviewer
- If no submissions in this list, show message as per design with link to all submissions page, no pre-filtering.
- There is no checkbox column and there are no batch actions.

**Acceptance criteria - as an AC reviewer**

- “Waiting for your review” will not link to the Submissions page when empty
- “Waiting for your review” will only populate with submissions in the external review status and assigned to the AC reviewer

#938 Previous Reviews on Dashboard
The block will be used on the dashboard to list submissions where the user has submitted a review for.

**Acceptance criteria**

- Title of block is 'Your previous reviews'
- Lists applications I have submitted a review for
- Applications ordered by submission date descending
- Table columns should be as per All Submissions tables
- Limit number of submissions to 5, if more than 5 show all link to appear (as per other summary tables) and link to all submissions page pre-filtered by reviewer
- If no submissions in this list, do not show block
- Filter and search available as per submissions summary block on submissions overview page
- Block shows for OTF staff and AC reviewers
- There are no batch actions available for this table

For OTF staff:

-  When I click through "Show All" under "Previous Reviews" block, I go to submissions/all filtered by Reviewers=myself

For AC reviewer:

- All submissions table is limited to only submissions I have reviewed. I cannot see any other submissions in this table regardless of the applied filters.
- There is no filter for "reviewed by"
- There is no filter for "reviewers"

### CODE REVIEW

**Comment by emullaney** [comment]
@todd-dembrey this is ready for review. It covers two tickets, 937 and 938. I think it will be easier to review on one PR because they are so connected. I've also merged this branch onto develop.

File: opentech/apply/funds/views.py [hunk 1]
```diff
@@ -147,13 +158,24 @@ def get_context_data(self, **kwargs):
         )
 
 
-class SubmissionListView(AllActivityContextMixin, BaseAdminSubmissionsTable, DelegateableListView):
+@method_decorator(staff_required, name='dispatch')
```
Comment (todd-dembrey) [inline opentech/apply/funds/views.py]: `ViewDispatcher` also handles the method decoration for you

File: opentech/apply/dashboard/templates/dashboard/reviewer_dashboard.html [hunk 2]
```diff
@@ -12,31 +16,64 @@ <h1 class="gamma heading heading--no-margin heading--bold">Dashboard</h1>
...
+
+            {% if display_more %}
+                <div class="all-submissions-table__more">
+                    <a href="{% url 'apply:submissions:list' %}">Show all</a>
+                </div>
             {% endif %}
-        </div>
+
+        {% else %}
+            <div class="grid grid--seventy-thirty wrapper wrapper--large ">
```
Comment (todd-dembrey) [inline opentech/apply/dashboard/templates/dashboard/reviewer_dashboard.html]: I dont think we want to put the seventy-thirty layout in just yet. we will not be adding the `30` component this block of sprints.

File: opentech/apply/dashboard/templates/dashboard/reviewer_dashboard.html [hunk 3]
```diff
@@ -1,6 +1,10 @@
 {% extends "base-apply.html" %}
 {% load render_table from django_tables2 %}
-{% load static %}
+{% load static wagtailcore_tags %}
```
Comment (todd-dembrey) [inline opentech/apply/dashboard/templates/dashboard/reviewer_dashboard.html]: Do we need the wagtail core tags here?

File: opentech/apply/dashboard/templates/dashboard/dashboard.html [hunk 4]
```diff
@@ -19,21 +23,65 @@ <h1 class="gamma heading heading--no-margin heading--bold">Dashboard</h1>
 <div class="wrapper wrapper--large wrapper--inner-space-medium">
         <div class="wrapper wrapper--large wrapper--inner-space-medium">
 
+            <h4 class="heading heading--normal">
+                Waiting for your review <span class="heading heading--submission-count">{{ in_review_count }}</span>
```
Comment (todd-dembrey) [inline opentech/apply/dashboard/templates/dashboard/dashboard.html]: Seeing as we use it across two pages, can we move this to an include?

File: opentech/apply/dashboard/templates/dashboard/dashboard.html [hunk 5]
```diff
@@ -19,21 +23,65 @@ <h1 class="gamma heading heading--no-margin heading--bold">Dashboard</h1>
...
-            <h3>Applications to review</h3>
-            {% if in_review.data %}
-                {% render_table in_review %}
-            {% else %}
-                No reviews to complete
-            {% endif %}
         </div>
+
+        <div class="wrapper wrapper--large wrapper--inner-space-medium">
+            <div class="wrapper wrapper--large wrapper--inner-space-medium">
```
Comment (chris-lawton) [inline opentech/apply/dashboard/templates/dashboard/dashboard.html]: Any particular reason for the nest divs with the same classes?

Comment (emullaney) [reply opentech/apply/dashboard/templates/dashboard/dashboard.html]: No reason, probably just an overzealous copy paste. I can adjust that!

File: opentech/apply/dashboard/tests/test_views.py [hunk 6]
```diff
@@ -61,3 +61,29 @@ def test_cannot_see_submission_in_determination_when_not_lead(self):
         ApplicationSubmissionFactory(status='concept_review_discussion', workflow_stages=2, form_data__title='Reviewr')
         response = self.get_page()
         self.assertNotContains(response, 'Reviewr')
+
+    def test_waiting_for_review_with_count(self):
+        submission = ApplicationSubmissionFactory(status='external_review', workflow_stages=2, reviewers=[self.user])
```
Comment (todd-dembrey) [inline opentech/apply/dashboard/tests/test_views.py]: Hmm, i think this is an error. Staff should not be reviewing during the external review phase. This is a preexisting issue with the view.

File: opentech/apply/dashboard/tests/test_views.py [hunk 7]
```diff
@@ -61,3 +61,29 @@ def test_cannot_see_submission_in_determination_when_not_lead(self):
...
+        response = self.get_page()
+        self.assertContains(response, 'Waiting for your review')
+        self.assertContains(response, submission.title)
+        self.assertEquals(response.context['in_review_count'], 1)
+
+    def test_no_submissions_waiting_for_review(self):
+        submission = ApplicationSubmissionFactory(status='external_review', workflow_stages=2, reviewers=[])
```
Comment (todd-dembrey) [inline opentech/apply/dashboard/tests/test_views.py]: need 1 more test to confirm they dont get submissions not in the external review phase.

File: opentech/apply/dashboard/views.py [hunk 8]
```diff
@@ -23,34 +26,88 @@ def get(self, request, *args, **kwargs):
...
 
 class ReviewerDashboardView(TemplateView):
+
     def get(self, request, *args, **kwargs):
         qs = ApplicationSubmission.objects.all().for_table(self.request.user)
 
-        my_review_qs = qs.in_review_for(request.user)
-        my_review = SubmissionsTable(my_review_qs, prefix='my-review-')
-        RequestConfig(request, paginate={'per_page': 10}).configure(my_review)
+        # Reviewer's current to-review submissions
```
Comment (todd-dembrey) [inline opentech/apply/dashboard/views.py]: can we pull some of this out into some methods just to tidy up what is going on.

File: opentech/apply/dashboard/views.py [hunk 9]
```diff
@@ -23,34 +26,88 @@ def get(self, request, *args, **kwargs):
...
+        # Replicating django_filters.views.FilterView
+        my_reviewed_qs = qs.reviewed_by(request.user).order_by('-submit_time')
+        kwargs = {
+            'data': self.request.GET or None,
+            'request': self.request,
+        }
+        kwargs.update({
+            'queryset': my_reviewed_qs,
+        })
+        self.filterset = SubmissionReviewerFilterAndSearch(**kwargs)
```
Comment (todd-dembrey) [inline opentech/apply/dashboard/views.py]: I dont think you need to assign queryset to `self`

File: opentech/apply/funds/models/submissions.py [hunk 10]
```diff
@@ -87,6 +87,10 @@ def in_review_for(self, user, assigned=True):
             qs = qs.filter(reviewers=user)
         return qs
 
+    def reviewed_by(self, user):
+        qs = self.filter(Q(reviews__author=user))
+        return qs
```
[the remark this answers is not in the document]
Comment (emullaney) [reply opentech/apply/funds/models/submissions.py]: Don't need Q!

File: opentech/apply/funds/views.py [hunk 11]
```diff
@@ -147,13 +158,24 @@ def get_context_data(self, **kwargs):
...
+class SubmissionAdminListView(AllActivityContextMixin, BaseAdminSubmissionsTable):
```
[the remark this answers is not in the document]
Comment (emullaney) [reply opentech/apply/funds/views.py]: Ah, I must have missed this when I was resolving conflicts with sprint-3, thanks!

File: opentech/apply/funds/views.py (hunk 1 above)
Comment (emullaney) [reply opentech/apply/funds/views.py]: yes, decorator is not needed, I've corrected, thanks!

File: opentech/apply/dashboard/templates/dashboard/reviewer_dashboard.html (hunk 2 above)
Comment (emullaney) [reply opentech/apply/dashboard/templates/dashboard/reviewer_dashboard.html]: I'll remove this css class definition entirely.

File: opentech/apply/dashboard/tests/test_views.py (hunk 6 above)
Comment (emullaney) [reply opentech/apply/dashboard/tests/test_views.py]: I can remove the test if it's a pre-existing issue... or try to fix if you tell me more about it. Let me know how to move forward.

**Comment by emullaney** [comment]
@todd-dembrey thanks for the review, this is ready for another look if you'd like.

File: opentech/apply/dashboard/tests/test_views.py (hunk 6 above)
Comment (todd-dembrey) [reply opentech/apply/dashboard/tests/test_views.py]: Lets leave it for now.

File: opentech/apply/dashboard/tests/test_views.py (hunk 7 above)
Comment (todd-dembrey) [reply opentech/apply/dashboard/tests/test_views.py]: @emullaney I'm not sure i can see a test case where the user is assigned as a reviewer but it isnt in the external phase.

**Comment by todd-dembrey** [comment]
Happy to assign over for testing, one small thing about the test otherwise great.

**Comment by emullaney** [comment]
@todd-dembrey, 
I've updated this test per your feedback and also made a fix that should hopefully solve the issue Maya caught with repeating submissions on the Waiting for your review block.

**Pull request merged**